File: c++/divide_and_conquer/q241.cpp
//divide and conquer
vector<int> diffWaysToCompute(string input) {
     vector<int> res;
     for(int i = 0; i < input.size(); i++){
         if(input[i] == '+' || input[i] == '-' || input[i] == '*'){
             vector<int> res1 = diffWaysToCompute(input.substr(0,i));
             vector<int> res2 = diffWaysToCompute(input.substr(i+1));

             for(int j = 0; j < res1.size(); j++){
                 for(int k = 0; k < res2.size(); k++){
                     if(input[i]=='-')
                         res.push_back(res1[j]-res2[k]);
                     else if(input[i]=='+')
                         res.push_back(res1[j]+res2[k]);
                     else
                         res.push_back(res1[j]*res2[k]);
                 }
             }
         }
     }
     //base case
     if(res.size()==0)
         res.push_back(atoi(input.c_str()));
     return res;

 }

File: python/divide_and_conquer/q241.py
"""
Given a string of numbers and operators, return all possible results from computing all the different possible ways to group numbers and operators. The valid operators are +, - and *.


Example 1
Input: "2-1-1".

((2-1)-1) = 0
(2-(1-1)) = 2
Output: [0, 2]


Example 2
Input: "2*3-4*5"

(2*(3-(4*5))) = -34
((2*3)-(4*5)) = -14
((2*(3-4))*5) = -10
(2*((3-4)*5)) = -10
(((2*3)-4)*5) = 10
Output: [-34, -14, -10, -10, 10]
"""
def diffWaysToCompute(self, input):
    """
    :type input: str
    :rtype: List[int]
    """
    if input.isdigit():
        return [int(input)]

    ret = []
    for i,val in enumerate(input):
        if val in "+-*":
            res1 = self.diffWaysToCompute(input[:i])
            res2 = self.diffWaysToCompute(input[i+1:])

            for val1 in res1:
                for val2 in res2:
                    if val == '+':
                        ret.append(val1+val2)
                    elif val == '-':
                        ret.append(val1-val2)
                    else:
                        ret.append(val1*val2)

    return ret
